lib/canister.rb

```diff
@@ -6,18 +6,49 @@
 # resolved at runtime. This allows for out-of-order declaration,
 # automatic dependency resolution, and--upon
 # redeclaration--automatic dependency cache invalidation.
+# @param hashlike [#each_pair, #keys #[]]
+# @see #fill_from_hashlike
 class Canister
-  def initialize
+  def initialize(hashlike = nil)
     @stack = []
     @registry = {}
     @resolved = {}
     @dependents = Hash.new do |hash, key|
       hash[key] = []
     end
     @mutex = Mutex.new
+    if hashlike
+      self.merge_from_hashlike(hashlike)
+    end
     yield self if block_given?
   end
 
+  # A "hashlike" is defined as anything that responds to both #[]
+  # and #keys and/or #each_pair
+  # Each value is wrapped in a `proc` and registered under the key.
+  # Note that Canister doesn't differentiate between symbols and
+  # strings for keys, so if your hashlike has keys of, e.g. both
+  # `"a"` and `:a` it won't work.
+  def merge_from_hashlike(hashlike)
+    iter = if hashlike.respond_to?(:each_pair)
+             hashlike.each_pair
+           else
+             if [:[], :keys].all? { |meth| hashlike.respond_to?(meth) }
+               hashlike.keys.lazy.map { |k| [k, hashlike[k]] }
+             else
+               msg = "Need something that responds to either #each_pair or both #keys and #[]"
+               raise ArgumentError.new(msg)
+             end
+           end
+    iter.each do |k, v|
+      blk = proc { v }
```

Comment:
What if `v` is a `Proc`? I can pass a proc like `canister.register(:foo, &my_proc)` but in a hashlike my proc is just going to be stored as a proc -- maybe not what I was expecting. This may be a can of worms re developer intent.